### DISCUSSION TITLE
Completed: Push to official docker hub

### DISCUSSION THREAD

**Discussion by lippoliv** [paperless-ngx/paperless-ngx#290] [Feature Requests] [answered]
Hi all,
today I switched my NAS from [linuxserver/paperless-ng](https://hub.docker.com/r/linuxserver/paperless-ng) to paperless-ngx. Because my Synology (currently) just supports official docker hub (or I did not research enough), I now [pushed paperless-ngx myself](https://hub.docker.com/r/lippertsweb/paperless-ngx).

Is there some plan to push images to official docker hub in the future?

**Comment by marcelbrueckner** [comment]
No issues pulling the images from the command-line on my DS1515+. No idea about the UI, though. I never use that. 

```bash
$ sudo docker pull ghcr.io/paperless-ngx/paperless-ngx:latest
latest: Pulling from paperless-ngx/paperless-ngx
f7a1c6dad281: Pull complete 
92c59ec44e08: Pull complete 
...
```

I currently run paperless-ng set up via Docker Compose. With the successful pull shown above I'm confident paperless-ngx will work as well.

However, for those wanting to use Docker's mirror registry feature, which only works with Docker Hub, having the images hosted on Docker Hub might be helpful. Setting up GitHub Actions to push to both registries should easily be done.

**Reply by lippoliv** [reply]
Ah sorry I was unprecise. 

Sure via CLI it will work ;) I was searching for a way via UI so SSH-unfamiliar users can easily so the project.

**Reply by sukisoft** [reply]
Would appreciate that as well. Though i am a developer myself, i dont really want to unneccessarily deal with complexity thats not needed in general. Using DSM also. did the setup via docker-compose via SSH to add all needed containers. worked well though.

**Comment by shamoon** [comment]
Ah, wasn't aware of this. @paperless-ngx/ci-cd is there any easy way we can also push to Docker hub without too much extra work / maintenance?

**Reply by schnuffle** [reply]
I think that's a easy task where the administrative work will take the most time.
When somebody sets up a account I can push to, it's as easy as using the buildandpush action with another repo

Regards

Schnuffle

**Comment by lippoliv** [comment]
I checked out myself how you could do that. 

### Option 1: make `build-docker-image` a matrix

This would be the easiest solution. Just make `build-docker-image` a matrix job with three inputs: Registry Host, -User and -Password. I don't like it really much, because build and inspect the image would be done twice, also the action would need to deal with the artifact, there should just be one.

### Option 2: push the image to multiple registries after build

This is some approach I like and basically is what I do right now on my machine. I use three lines of bash to achieve that

```shell
docker pull ghcr.io/paperless-ngx/paperless-ngx:$1
docker tag ghcr.io/paperless-ngx/paperless-ngx:$1 lippertsweb/paperless-ngx:$1
docker push lippertsweb/paperless-ngx:$1
```

I saw in GitHub actions the usage of [docker/build-push-action](https://github.com/docker/build-push-action), they have an documentation on [how to copy images between registries](https://github.com/docker/build-push-action/blob/master/docs/advanced/copy-between-registries.md).

The key challenge I see here is, that `${{ steps.prepare.outputs.tags }}` holds multiple tags to push to.

**Reply by schnuffle** [reply]
${{ steps.prepare.outputs.tags }}

As we only want final version to be pushed to dockerhub, it'S easy to adopt.

Regards

Schnuffle

**Reply by shamoon** [reply]
Ok great I’ll setup an account for the org and ping you back when it’s ready. Thanks!

**Reply by sytone** [reply]
I am pushing a container to ghcr.io, shout if you need help in setting up the action as I just did this in the last week myself.

**Comment by Schneppi** [comment]
> Hi all, today I switched my NAS from [linuxserver/paperless-ng](https://hub.docker.com/r/linuxserver/paperless-ng) to paperless-ngx. Because my Synology (currently) just supports official docker hub (or I did not research enough), I now [pushed paperless-ngx myself](https://hub.docker.com/r/lippertsweb/paperless-ngx).
> 
> Is there some plan to push images to official docker hub in the future?

Does it work? Because i tried your build on my Syno (adding only PAPERLESS_ADMIN_USER and PAPERLESS_ADMIN_PASSWORD env.variables). I can log in but if i try to upload a pdf i get http 500 errors...

**Reply by lippoliv** [reply]
Yes I do run successfully in Synology. Installation with >2000 documents. 

You need to spend some time with permissions probably. Do you follow some tutorial?

**Reply by Schneppi** [reply]
No. I've just installed it using Synology's UI.

**Reply by Schneppi** [reply]
> You need to spend some time with permissions probably. Do you follow some tutorial?

Do you know of any Tutorial you can recommend? Thank you :)

**Reply by littletruckman** [reply]
Maybe you need to adjust / add the following environment variables for your webserver container so they match the UID and GID of the user used to create the shared folders you probably created on your synology to make your persistant storages (media and consume folders)

```
# The UID and GID of the user used to run paperless in the container. Set this
# to your UID and GID on the host so that you have write access to the
# consumption directory.
      USERMAP_UID: 10xx
      USERMAP_GID: 10y
```
 
A good post to do identify these values is from Marius Bogdan Lixandru (Marius Hosting) 
https://mariushosting.com/synology-how-to-find-uid-userid-and-gid-groupid/ 

Hope it help :-)

**Reply by Schneppi** [reply]
Thank you. I am sure this will help. But another issue is that the Docker Machine does not create the necessary Folders on the NAS. And i was not able to successfully change this by using environment variables.

I think this Docker Repo is not of any use for Synologys which had never used a Paperless Docker before.

**Reply by lippoliv** [reply]
@Schneppi the folders you need to create yourself. Then you need to set permissions (at least in my case).

I don't have a link to some tutorial yet. I have locally a documentation what to do. I could submit something here. But I don't know what would be the best place?

**Reply by marcogiorgio** [reply]
> @Schneppi the folders you need to create yourself. Then you need to set permissions (at least in my case).
> 
> I don't have a link to some tutorial yet. I have locally a documentation what to do. I could submit something here. But I don't know what would be the best place?



> Thank you. I am sure this will help. But another issue is that the Docker Machine does not create the necessary Folders on the NAS. And i was not able to successfully change this by using environment variables.
> 
> I think this Docker Repo is not of any use for Synologys which had never used a Paperless Docker before.

Hi,
this is what I've done in order to make it work on my Synology, hope it will help:

1. Create a user named "**paperless**" from the Synology control panel
2. Get the user id and the guid of the user just created, by using the command: "**id paperless**" via SSH
3. Create the folders: 

    - **/volume1/docker/Paperless/data**
    - **/volume1/docker/Paperless/media**

4. Give "paperless" user RO permission to the folder /volume1/docker from the Synology control panel
5. Give "paperless" user RW permissions to the folder  /volume1/docker/Paperless and its subfolders
6. In the docker-compose.yml set the env variables **USERMAP_UID, USERMAP_GID** with the values obtained from point 2
7. In the docker-compose file use the bind mounts:

    -  **/volume1/docker/Paperless/data:/usr/src/paperless/data**
    - **/volume1/docker/Paperless/media:/usr/src/paperless/media**

I've not specified how to configure the DB because I am not aware of your full setup.

**Reply by Schneppi** [reply]
@marcogiorgio @lippoliv @littletruckman
Thank you all for your kind help. I currently have very little free time and can't get around to doing more testing on the Synology.
I've deceided to install Paperless-ngx on a Raspberry Pi 4B later this year.

**Reply by KBischof** [reply]
Just my two cents, check out this tutorial for installing paperless-ngx on Synology. I have not tried myself, so cannot vouch if it works or if it even is the preferred method for installing on Synology NAS. 

https://mariushosting.com/how-to-install-paperless-ngx-on-your-synology-nas/

**Reply by Schneppi** [reply]
> Just my two cents, check out this tutorial for installing paperless-ngx on Synology. I have not tried myself, so cannot vouch if it works or if it even is the preferred method for installing on Synology NAS. 
> 
> 
> 
> https://mariushosting.com/how-to-install-paperless-ngx-on-your-synology-nas/
> 
> 

Awesome! Thank you!

**Comment by ashrude** [comment]
Looks like linuxserver released a paperless-ngx version

**Comment by stumpylog** [comment, marked as answer]
We're now pushing our own image to the Docker Hub.  Currently, this is only development images, but after our next release all releases will be available from either Docker Hub or GHCR

**Reply by shamoon** [reply]
Haha we answered at the same time, yes nice work!

https://hub.docker.com/r/paperlessngx/paperless-ngx
